goben/prober.go

```diff
@@ -0,0 +1,289 @@
+	probeInterval time.Duration // probe probeInterval
+	pktInterval	  time.Duration // packet sending interval
+	pktsPerProbe  int           // number of packets sent per probe
+}
+
+type Prober struct {
+	config		ProberConfig
+	conn		*icmp.PacketConn	// the ICMP connection
+	runCnt		uint64				// a counter that helps to construct seq# and runId
+	fileLogger  *log.Logger			// logger that will log message to a file
+}
+
+// Initialize a new Prober instance
+// Need to be called before any of the following
+func (p *Prober) Init(config ProberConfig) error {
+	p.config = config
+
+	// prepare logger
+	filePath := fmt.Sprintf("./stats_%v.txt", config.targets[0]) // todo: the file name could be a user input
+	file, err := os.Create(filePath)
+	if err != nil {
+		log.Panicf("Cannot create a logging file to persist network traffic statistics! %v\n", err.Error())
+	}
+	p.fileLogger = log.New(file, "", log.LstdFlags)
+
+	return p.listen()
+}
+
+// create a new icmp connection to listen to in coming packets
+func (p *Prober) listen() error {
+	opts := p.config
+	if opts.proto == "" {
+		log.Fatalf("The prober from host %s misses configuration info.\n", opts.source)
+	}
+	var err error
+	p.conn, err = icmp.ListenPacket(opts.proto, "0.0.0.0") // use 0.0.0.0 here meaning we listen to any packets regardless if the packet is addressed to myself
+	return err
+}
+
+// construct the ICMP message and marshall it to bytes
+func (p *Prober) packetToSend(runId, seq uint16) []byte {
+	// todo: handle UDP
+	msg := &icmp.Message{
+		Type: ipv4.ICMPTypeEcho,
+		Code: 0,
+		Body: &icmp.Echo{
+			ID: int(runId),
+			Seq: int(seq),
+			Data: timeToBytes(time.Now()),
+		},
+	}
+	bytes, err := msg.Marshal(nil)
+	if err != nil {
+		// This should never happen
+		log.Panicf("Error marshalling the ICMP message. Err: %v\n", err)
+	}
+	return bytes
+}
+
+// unmarshall the bytes back to ICMP message
+func (p *Prober) packetToRecv(pktbuf []byte) (net.IP, *icmp.Message, error) {
+	n, sender, err := p.conn.ReadFrom(pktbuf)
+	if err != nil {
+		return nil, nil, err
+	}
+	// get the sender's IP address
+	// Since sender is an interface net.Addr, we have to cast it down to net.IPAddr/net.UDPAddr type to get the IP
+	var senderIP net.IP
+	switch sender := sender.(type) {
+	case *net.IPAddr:
+		senderIP = sender.IP
+	case *net.UDPAddr:
+		senderIP = sender.IP
+	}
+
+	msg, err := icmp.ParseMessage(protocolICMP, pktbuf[:n])
+	if err != nil {
+		return nil, nil, err
+	}
+	return senderIP, msg, nil
+}
+
+// send ICMP message
+// We sent the icmp packets synchronously
+func (p *Prober) send(runID uint16, morePkts chan bool) {
+	seq := runID & uint16(0xff00)
+	for i := 0; i < p.config.pktsPerProbe; i++ {
+		for _, target := range p.config.targets {
+			log.Printf("Request to=%s id=%d seq=%d", target, runID, seq)
+			if _, err := p.conn.WriteTo(p.packetToSend(runID, seq), parseIP(target)); err != nil {
+				log.Println(err.Error())
+				continue
+			}
+			fmt.Println("SEND A ICMP PACKET")
```

Comment:
This is leftover debug code. Should be removed.

Reply:
done